### ISSUE TITLE
CA1859: Nonsensical false positive in VB.NET iterator method

### ISSUE THREAD

**Issue by reduckted** [dotnet/roslyn-analyzers#7029]
### Analyzer

**Diagnostic ID**: [CA1859](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1859): `Use concrete types when possible for improved performance`

### Analyzer source

**SDK**: Built-in CA analyzers in .NET 8 SDK (RC2)

**Version**: 8.0.100-rc.2.23502.2

### Describe the bug

Under certain conditions, an `Iterator` method in VB.NET will produce CA1859 with a suggestion to change the type to something that makes no sense at all.

### Steps To Reproduce

Consider this code:
```vb
Public NotInheritable Class Test
    Private Shared Iterator Function CreateWidgets(definitions As IEnumerable(Of KeyValuePair(Of String, Integer))) As IEnumerable(Of Widget)
        For Each item In definitions.OrderBy(Function(x) x.Key)
            Yield New Widget
        Next item
    End Function
End Class

Public Class Widget
End Class
```

### Expected behavior

No diagnostic should be produced.

### Actual behavior

The diagnostic is produced with the message:

> Change return type of method 'CreateWidgets' from 'IEnumerable(Of Widget)' to 'String' for improved performance

## Additional context

This seems to be related to the `OrderBy` call. Just using this for loop does not produce the diagnostic:

```vb
For Each item In definitions
```

Neither does this:
```vb
For Each item In definitions.OrderBy(Function(x) 0)
```

But this does:
```vb
For Each item In definitions.OrderBy(Function(x) "")
```

An even simpler example that incorrectly produces the diagnostic is this:
```vb
Private Shared Iterator Function Create() As IEnumerable(Of Integer)
    For Each item In {1}.OrderBy(Function(x) "")
        Yield 1
    Next item
End Function

' Produces:
'     CA1859 Change return type of method 'Create' from 'IEnumerable(Of Integer)' to 'String' for improved performance	
```

**Comment by geeknoid** [comment]
Related to #6845, the rule just doesn't have the language-specific knowledge needed to handle these constructs property.
